### PR TITLE
[REVIEW] Port unary to libcudf++

### PR DESCRIPTION
**Pull request by codereport**
[ISSUE] 2950

Port math_ops.cu
Port unary_ops.cuh

---

null_ops was done in: 3182
cast_ops was done in: 3397

### CODE REVIEW

**Comment by harrism** [comment]
@codereport please avoid force pushes (e.g. use merge rather than rebase) once you have an open PR associated with a feature branch.  Github doesn't handle force pushes well. It's most important once people have reviewed the code, because force pushing disassociates comments from the current state of the code.

**Comment by codereport** [comment]
> please avoid force pushes

Will avoid these in the future! Thanks!

**Comment by harrism** [comment]
@jrhemstad Unary ops encompasses a lot more than just null_ops! But yes, null_ops is a *new* feature added by @rgsl888prabhu in #2962 and immediately ported to libcudf in #3182.  @codereport I don't think you need to touch null_ops, just everything else in unary.hpp

**Comment by codereport** [comment]
@jrhemstad @harrism : @rgsl888prabhu did mention he had code changes for `is_null` and `is_not_null` but I though we was just letting me know about the pre-ported changes he merged. Probably could have been avoided if I was notified of the PR numbers (if they aren't already merged that is). 

I will focus on the 3 other files.

**Comment by harrism** [comment]
null_ops did not exist when I originally created #2950.

**Comment by codereport** [comment]
@harrism Sorry Mark! I wasn't clear - I meant that if Ram had have let me know what the PR #'s were I could have checked. I will just make sure in the future to ask for them :)

**Comment by rgsl888prabhu** [comment]
I apologize for any inconvenience caused by me.

**Comment by harrism** [comment]
GitHub has a pretty good PR search function. :)

**Comment by jrhemstad** [comment]
@codereport just as a general note, I think the old implementation could be vastly improved by replacing the custom kernel with a `thrust::transform`.

**Comment by codereport** [comment]
> thrust::transform

Thanks @jrhemstad! I will definitely do that

File: cpp/src/unary/math_ops.cu [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    typename std::enable_if_t<!is_supported<T>(), void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+} // namespace detail
+
+std::unique_ptr<cudf::column> 
+unary_operation(cudf::column_view input,
```
Comment (harrism) [inline cpp/src/unary/math_ops.cu]: See https://github.com/rapidsai/cudf/blob/cd3cfc5d68e964246881a44602d5a299959ac518/cpp/docs/TRANSITIONGUIDE.md#inputoutput-style

```suggestion
unary_operation(cudf::column_view const& input, 
```

File: cpp/src/unary/math_ops.cu [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    template <typename T>
+    typename std::enable_if_t<!is_supported<T>(), void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+} // namespace detail
+
+std::unique_ptr<cudf::column>
```
Comment (harrism) [inline cpp/src/unary/math_ops.cu]: Shouldn't the headers for these files be part of this PR also?

File: cpp/src/unary/unary_ops.cuh [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            static_cast<Tout*>(device_output->head()),
+            functor
+        );
+
+        CUDA_CHECK_LAST();
+        return GDF_SUCCESS;
+    }
+};
+
+inline void handleChecksAndValidity(column_view input, mutable_column_view output) {
```
Comment (harrism) [inline cpp/src/unary/unary_ops.cuh]:
```suggestion
inline void handleChecksAndValidity(column_view const& input, mutable_column_view const& output) {
```

File: cpp/src/unary/unary_ops.cuh [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    for (int i = start; i < size; i += step) {
+        results[i] = functor.apply(data[i]);
+    }
+}
+
+template<typename T, typename Tout, typename F>
+struct Launcher {
+    static
+    gdf_error launch(cudf::column_view input,
```
Comment (harrism) [inline cpp/src/unary/unary_ops.cuh]:
```suggestion
    gdf_error launch(cudf::column_view const& input,
```

File: cpp/src/unary/unary_ops.cuh [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                     cudf::mutable_column_view output) {
```
Comment (harrism) [inline cpp/src/unary/unary_ops.cuh]:
```suggestion
                     cudf::mutable_column_view & output) {
```

File: cpp/src/unary/unary_ops.cuh (hunk 3 above)
Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: should be & on the mutable, will update

File: cpp/src/unary/math_ops.cu (hunk 2 above)
Comment (codereport) [reply cpp/src/unary/math_ops.cu]: yes, will be added those

File: cpp/src/unary/unary_ops.cuh (hunk 5 above)
Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: will update without the space

File: cpp/src/unary/unary_ops.cuh (hunk 4 above)
Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: will update

File: cpp/src/unary/math_ops.cu [hunk 6]
```diff
@@ -35,47 +35,47 @@ namespace detail {
 struct DeviceSin {
     template<typename T>
     __device__
-    T apply(T data) {
+    T operator()(T data) {
         return std::sin(data);
```
Comment (jrhemstad) [inline cpp/src/unary/math_ops.cu]: @harrism currently these are using the `std` operators, which works in device code because they are `constexpr`.

What are your thoughts on if we should instead be using the versions of these operators from the CUDA math lib? https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html

Comment (harrism) [reply cpp/src/unary/math_ops.cu]: I'm guessing that in device code that std::sin resolves to the CUDA math library's sin() anyway? Otherwise how would it work?

Comment (harrism) [reply cpp/src/unary/math_ops.cu]: I think the advantage of the `std` versions is that there are overloads for `float` and `double` so we can get the most efficient code without explicitly calling `sinf`, for example.

Comment (jrhemstad) [reply cpp/src/unary/math_ops.cu]: Yeah, looks like they're identical: https://godbolt.org/z/T4fN_i

**Comment by harrism** [comment]
@codereport I saw you added cast ops to this PR. @trxcllnt needed  them so he went ahead and implemented that for libcudf++: #3397

**Comment by codereport** [comment]
@harrism yep @trxcllnt discussed it with me a couple weeks ago.

File: cpp/src/unary/math_ops.cu [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+struct DeviceSin {
+    template<typename T,
+             typename std::enable_if_t<!std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
+    __device__
+    T operator()(T data) {
+        return std::sin(data);
+    }
+
+    template <typename T,
+              typename std::enable_if_t<std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
```
Comment (jrhemstad) [inline cpp/src/unary/math_ops.cu]: This special code path for `bool8` seems annoying to have to maintain. I think it would be unnecessary if `bool8` had an implicit conversion operator to arithmetic types.

Comment (codereport) [reply cpp/src/unary/math_ops.cu]: That was one of the first things that I tried (adding an implicit / explicit conversion to the `bool8` type) - however it didn't seem to work. I discussed with @trxcllnt the summary from out chat was:
> oh so looking back at the error you're getting... std::sin only has overloads for double, float, and long double? so IIUC the compiler wouldn't automatically call the uint8_t conversion operator.

Comment (codereport) [reply cpp/src/unary/math_ops.cu]: However - if I am missing something and there is a way to do this - I agree it would be much nicer. Having this duplicate code in each function object is very irritating.

Comment (jrhemstad) [reply cpp/src/unary/math_ops.cu]: I think something like this should work:
```
template <typename T, std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
operator T ();
```

The compiler should select the conversion to `float/double`.

Comment (codereport) [reply cpp/src/unary/math_ops.cu]: I gave a couple variants of this a shot:
```
  template <typename T, std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
  CUDA_HOST_DEVICE_CALLABLE operator T() {
    return static_cast<T>(static_cast<bool>(this->value));
  }
```
but no dice. I had seen your post to @harrism earlier as well:
> Yeah, basically add what the old bool8 had: https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/include/cudf/utilities/legacy/wrapper_types.hpp#L107
and tried something similar. 

Did you have something different in mind for the body of the conversion operator? I am not a TMP expert so I might be missing something simple. I will take a stab at this tmo again

File: CHANGELOG.md [hunk 8]
```diff
@@ -181,7 +179,7 @@
 - PR #3386 Removing external includes from `column_view.hpp`
 - PR #3369 Add write_partition to dask_cudf to fix to_parquet bug
 - PR #3388 Support getitem with bools when DataFrame has a MultiIndex
-- PR #3408 Fix String and Column (De-)Serialization
```
Comment (harrism) [inline CHANGELOG.md]: Why did you delete #3408?

File: cpp/src/unary/unary_ops.cuh [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                          input.begin<T>(),
+                          input.end<T>(),
+                          output.begin<Tout>(),
+                          F{});
+
+        CUDA_CHECK_LAST();
+    }
+};
+
+inline void handleChecksAndValidity(column_view const& input, mutable_column_view& output) {
```
Comment (harrism) [inline cpp/src/unary/unary_ops.cuh]: Not a fan of this multi-purpose function... 

```suggestion
inline void handle_checks_and_validity(column_view const& input, mutable_column_view& output) {
```

File: cpp/src/unary/unary_ops.cuh [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+#include <cudf/cudf.h>
+#include <cudf/column/column_view.hpp>
+#include <cudf/null_mask.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace unary {
+
+template<typename T, typename Tout, typename F>
+struct Launcher {
```
Comment (harrism) [inline cpp/src/unary/unary_ops.cuh]:
```suggestion
struct launcher {
```

File: cpp/src/unary/math_ops.cu (hunk 7 above)
Comment (jrhemstad) [reply cpp/src/unary/math_ops.cu]: That looks right to me. What didn't work?

File: CHANGELOG.md (hunk 8 above)
Comment (codereport) [reply CHANGELOG.md]: sorry, was using a new mergetool when i was offline and clearly messed up the CHANGELOG.md, will fix

File: cpp/src/unary/unary_ops.cuh (hunk 10 above)
Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: Done.

File: cpp/src/unary/unary_ops.cuh (hunk 9 above)
Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: Done (updating name). 

@jrhemstad, can I confirm we need this function? I asked you before and you said yes, but it looks like @trxcllnt removed it when porting unary cast in 3397

File: cpp/src/unary/math_ops.cu (hunk 7 above)
Comment (codereport) [reply cpp/src/unary/math_ops.cu]: You get the same error as before:
```
/home/choekstra/ws3/cudf/cpp/src/unary/math_ops.cu(38): error: more than one instance of overloaded function "std::sin" matches the argument list:
            function "sin(double)"
            function "std::sin(float)"
            function "std::sin(long double)"
            argument types are: (cudf::experimental::bool8)
```

**Comment by harrism** [comment]
@codereport is this ready for review? If so, please mark it so and we can do another review pass.

**Comment by codereport** [comment]
@harrism I am just in the process of adding a couple extra tests, will mark as [REVIEW] once I have added these.

File: cpp/tests/unary/legacy/unary_ops_test.cu [hunk 11]
```diff
@@ -2383,16 +2383,18 @@ TYPED_TEST(gdf_logical_test, LogicalNot) {
...
+		});
 
 	// Use vector to build expected output
-	auto expectCol = cudf::test::column_wrapper<cudf::bool8>{h_expect_v};
-
-	auto output = cudf::unary_operation(inputCol, cudf::unary_op::NOT);
-
+	auto output    = cudf::unary_operation(inputCol, cudf::unary_op::NOT);
 	auto outputCol = cudf::test::column_wrapper<cudf::bool8>(output);
+	auto expectCol = cudf::test::column_wrapper<cudf::bool8>{h_expect_v};
```
Comment (jrhemstad) [inline cpp/tests/unary/legacy/unary_ops_test.cu]: `cudf::bool8` is the old one. Use `cudf::experimental::bool8`.

Comment (rgsl888prabhu) [reply cpp/tests/unary/legacy/unary_ops_test.cu]: @jrhemstad but this test is legacy.

File: cpp/src/unary/unary_ops.cuh [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                          input.begin<T>(),
+                          input.end<T>(),
+                          output.begin<Tout>(),
+                          F{});
+
+        CUDA_CHECK_LAST();
+    }
+};
+
+inline void handle_checks_and_validity(column_view const& input, mutable_column_view& output) {
```
Comment (trxcllnt) [inline cpp/src/unary/unary_ops.cuh]: Can this function be removed by using `copy_bitmask` and constructing the output column directly, like I am here?
https://github.com/rapidsai/cudf/blob/189e68adc3b3f8ba0c2a47faa9d1785e1da45dc0/cpp/src/unary/cast_ops.cu#L67-L70

File: cpp/tests/unary/legacy/unary_ops_test.cu (hunk 11 above)
Comment (codereport) [reply cpp/tests/unary/legacy/unary_ops_test.cu]: Yea, sorry. This was my OCD not liking the raw for loop :)

File: cpp/src/unary/unary_ops.cuh (hunk 12 above)
Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: I will take a look rn and let you know. Thanks @trxcllnt

Comment (codereport) [reply cpp/src/unary/unary_ops.cuh]: Yep! I was able to make it work! Thanks for your help / the suggestion!

File: CHANGELOG.md [hunk 13]
```diff
@@ -146,6 +146,7 @@
 - PR #3425 Strings column copy_if_else implementation
 - PR #3422 Move utilities to legacy
 - PR #3201 Define and implement new datetime_ops APIs
+- PR #3214 Port unary to libcudf++
```
Comment (harrism) [inline CHANGELOG.md]: Match other libcudf++ port changelogs ("libcudf++" name isn't helpful externally)

```suggestion
- PR #3214 Define and implement new unary operations APIs
```

File: cpp/src/unary/math_ops.cu (hunk 7 above)
Comment (harrism) [reply cpp/src/unary/math_ops.cu]: How about factoring out the casting so that all operators can just call a single function rather than maintaining two copies of every operator? Something that calls the actual op with casts only for types that need it, like this. Could be part of a base class that all device_ops inherit.

```
template<typename T, typename Op, std::enable_if_t<!std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
__device__ T normalized_unary_op()(Op op, T data) { return Op(data); }

template<typename T, std::enable_if_t<std::is_same<T, cudf::experimental::bool8>::value>* = nullptr>
__device__ T normalized_unary_op()(T data) { return static_cast<T>(Op(static_cast<float>(data)); }
```

File: cpp/tests/unary/unary_ops_test.cu [hunk 14]
```diff
@@ -55,6 +55,155 @@ cudf::test::fixed_width_column_wrapper<T> create_expected_columns(cudf::size_typ
...
+    auto output = cudf::experimental::unary_operation(input, cudf::experimental::unary_op::ABS);
+
+    cudf::test::expect_columns_equal(expected, output->view());
+}
+
+TYPED_TEST(cudf_math_test, SQRT)
+{
+    using T = TypeParam;
+
+    cudf::size_type colSize = 175;
```
Comment (harrism) [inline cpp/tests/unary/unary_ops_test.cu]: Please use a size that spans multiple thread blocks.

File: cpp/tests/unary/unary_ops_test.cu [hunk 15]
```diff
@@ -55,6 +55,155 @@ cudf::test::fixed_width_column_wrapper<T> create_expected_columns(cudf::size_typ
...
+}
+
+TYPED_TEST(cudf_math_with_floating_point_test, SimpleCEIL)
+{
+    cudf::test::fixed_width_column_wrapper<TypeParam> input    {{ 1.1, 3.3, 5.5, 7.7 }};
+    cudf::test::fixed_width_column_wrapper<TypeParam> expected {{ 2.0, 4.0, 6.0, 8.0 }};
+    auto output = cudf::experimental::unary_operation(input, cudf::experimental::unary_op::CEIL);
+    cudf::test::expect_columns_equal(expected, output->view());
+}
+
```
Comment (harrism) [inline cpp/tests/unary/unary_ops_test.cu]: I think we need 
* test of unary ops on error cases (which things should throw exceptions?)
* test of unary ops on empty inputs (result should be empty outputs of appropriate type).

File: cpp/tests/unary/unary_ops_test.cu [hunk 16]
```diff
@@ -55,6 +55,155 @@ cudf::test::fixed_width_column_wrapper<T> create_expected_columns(cudf::size_typ
...
+    auto output = cudf::experimental::unary_operation(input, cudf::experimental::unary_op::NOT);
+    cudf::test::expect_columns_equal(expected, output->view());
+}
+
+template <typename T>
+struct cudf_math_test : public cudf::test::BaseFixture {};
+
+TYPED_TEST_CASE(cudf_math_test, cudf::test::NumericTypes);
+
+TYPED_TEST(cudf_math_test, ABS)
```
Comment (harrism) [inline cpp/tests/unary/unary_ops_test.cu]: Curious, why are ABS and SQRT are tested differently than all other ops?

Comment (codereport) [reply cpp/tests/unary/unary_ops_test.cu]: These functions (with the input I set up) don't involve truncation - if I used others I might get a rounding error and for some the functions fail compilation due `narrowing conversion`

File: cpp/tests/unary/unary_ops_test.cu (hunk 14 above)
Comment (codereport) [reply cpp/tests/unary/unary_ops_test.cu]: Done.

File: CHANGELOG.md (hunk 13 above)
Comment (codereport) [reply CHANGELOG.md]: Done

File: cpp/include/cudf/unary.hpp [hunk 17]
```diff
@@ -22,6 +22,37 @@
...
+ * @brief Performs unary op on all values in column
+ *
+ * @param[in] input A `column_view` as input
+ * @param[in] op operation to perform
+ *
+ * @returns std::unique_ptr<cudf::column> Result of the operation
+ */
+std::unique_ptr<cudf::column> unary_operation(cudf::column_view const& input,
+                                              cudf::experimental::unary_op op,
+                                              cudaStream_t stream = 0,
```
Comment (jrhemstad) [inline cpp/include/cudf/unary.hpp]: Stream doesn't go in public API.

File: cpp/src/unary/math_ops.cu (hunk 7 above)
Comment (jrhemstad) [reply cpp/src/unary/math_ops.cu]: This needs to be resolved in the `bool8` type and not externally in functions like this. I'm not sure what the right combinations of implicit/explicit conversion and operators is necessary, but it's not sustainable to have to do things like what was done here in providing two versions of every operator.

File: cpp/tests/unary/unary_ops_test.cu (hunk 15 above)
Comment (codereport) [reply cpp/tests/unary/unary_ops_test.cu]: `normalized_unary_op` refactor done (and I very good suggestion I might add - I am sad I didn't realize this myself).

File: cpp/src/unary/math_ops.cu [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+                input.null_count());
+
+        } else {
+            return cudf::experimental::allocate_like(input);
+        }
+    } ();
+
+    if (input.size() == 0) return output;
+
+    auto output_view = output->mutable_view();
```
Comment (jrhemstad) [inline cpp/src/unary/math_ops.cu]: This pattern of passing in the output as a mutable_view prevents any future support of string or other non fixed-width type columns. Granted, most of these unary ops don't make sense on strings. However, I believe these should be modified to _return_ their output rather than pass via a mutable view.

File: cpp/include/cudf/unary.hpp (hunk 17 above)
Comment (codereport) [reply cpp/include/cudf/unary.hpp]: Good catch, fixed.

File: cpp/src/unary/math_ops.cu (hunk 18 above)
Comment (codereport) [reply cpp/src/unary/math_ops.cu]: dispatcher / launcher is updated to return `std::unique_ptr<cudf::column_view>`

File: cpp/tests/unary/unary_ops_test.cu (hunk 15 above)
Comment (codereport) [reply cpp/tests/unary/unary_ops_test.cu]: failures / throwing tests added as well
